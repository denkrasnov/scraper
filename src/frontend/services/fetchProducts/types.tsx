import { ProductName, Article } from "../../../backend/scrapers/types";

export enum ActionTypes {
  FETCH = "FETCH",
  FETCH_INIT = "FETCH_INIT",
  FETCH_ERROR = "FETCH_ERROR",
  FETCH_SUCCESS = "FETCH_SUCCESS"
}

export interface State {
  isLoading: boolean;
  isError: string;
  query: string;
  products: Products | null;
}

export interface Fetch {
  type: ActionTypes.FETCH;
  payload: string;
}

interface ActionSuccess {
  type: ActionTypes.FETCH_SUCCESS;
  payload: Products;
}

interface ActionInit {
  type: ActionTypes.FETCH_INIT;
}

interface ActionError {
  type: ActionTypes.FETCH_ERROR;
  payload: any;
}

export type Action = ActionSuccess | ActionInit | ActionError | Fetch;

export interface Products {
  name: ProductName;
  items: Article[];
}

export interface Response {
  data: Products;
}
