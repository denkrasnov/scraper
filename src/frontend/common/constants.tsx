import { Channel } from "../../backend/scrapers/types";

export const channelLogo: Record<string, string> = {
  [Channel.JurnalTV]:
    "https://live-tv-channels.org/pt-data/uploads/logo/md-jurnal-tv.jpg",
  [Channel.NTV]:
    "https://yt3.ggpht.com/ytc/AAUvwnjTxNZwMTMTlj18m75pswpomJFvJEHuargB5gWA=s900-c-k-c0x00ffffff-no-rj",
  [Channel.TV8]: "https://tv8.md/wp-content/uploads/2020/07/logo-tv8-new-2.png"
};
