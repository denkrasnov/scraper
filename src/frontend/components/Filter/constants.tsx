// TODO: created shared types with /backend
import { ProductName } from "../../../backend/scrapers/types";

export const filterName = {
  [ProductName.MD]: "Canal"
};
