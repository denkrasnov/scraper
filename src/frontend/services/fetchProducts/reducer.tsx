import { Action, State, ActionTypes } from "./types";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.FETCH:
      return {
        ...state,
        query: action.payload,
        isError: ""
      };
    case ActionTypes.FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        products: action.payload
      };
    case ActionTypes.FETCH_ERROR:
      return {
        ...state,
        isLoading: false,
        isError: action.payload
      };

    default:
      throw new Error();
  }
};

export default reducer;
