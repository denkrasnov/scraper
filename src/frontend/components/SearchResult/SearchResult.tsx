import React, { FC } from "react";
import { useQuery, gql } from "@apollo/client";

import { Article, ProductName, Channel } from "../../../backend/scrapers/types";
import Box from "~app/atoms/Box";
import Loader from "~app/atoms/Loader";
import Grid from "~app/atoms/Grid";
import ProductList from "../ProductList";
import Filter from "../Filter";

const FETCH_NEWS = gql`
  query FetchNews {
    news {
      _id
      date
      header
      imageURL
      newsURL
      channel
    }
  }
`;

const SearchResult: FC = () => {
  const { loading, error, data } = useQuery<{ news: Article[] }>(FETCH_NEWS);

  if (loading || error || !data) {
    return (
      <Box justifyContent="center" marginTop="s48">
        {loading ? <Loader /> : "Error..."}
      </Box>
    );
  }

  const { news } = data;

  return (
    <Grid result>
      <Filter
        items={news}
        options={[Channel.NTV, Channel.TV8, Channel.JurnalTV]}
        productName={ProductName.MD}
      >
        <ProductList />
      </Filter>
    </Grid>
  );
};

export default SearchResult;
