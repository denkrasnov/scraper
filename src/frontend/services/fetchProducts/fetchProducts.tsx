import { useEffect, useReducer, Dispatch } from "react";
import axios from "axios";

import { INITIAL_STATE } from "./constants";
import reducer from "./reducer";
import { Action, State, ActionTypes, Response } from "./types";

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });
      try {
        const result: Response = await axios.get(`/search/${query}`);

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

export default fetchProducts;
