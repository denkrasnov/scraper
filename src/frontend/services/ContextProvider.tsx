import React, { FC, useContext, createContext } from "react";

import { INITIAL_STATE } from "./fetchProducts/constants";
import fetchProducts from "./fetchProducts";
import { ContextProviderProps, CreateContext } from "./types";

export const StateContext = createContext<CreateContext>([
  INITIAL_STATE,
  () => {}
]);

const ContextProvider: FC<ContextProviderProps> = ({ children }) => (
  <StateContext.Provider value={fetchProducts()}>
    {children}
  </StateContext.Provider>
);

export const useFullContext = () => useContext(StateContext);

export default ContextProvider;
