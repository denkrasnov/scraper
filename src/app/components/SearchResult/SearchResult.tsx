import React, { FC } from "react";

import Box from "src/atoms/Box";
import { useFullContext } from "src/app/services/fetchProducts";
import ProductCard from "../ProductCard";

const SearchResult: FC = () => {
  const {
    products: [{ products, isLoading, isError }]
  } = useFullContext();

  console.log("products---->", products);

  if (isLoading || isError) {
    return isLoading ? <Box>Loading...</Box> : <Box>Error...</Box>;
  }

  return (
    <Box flexWrap="wrap">
      {products &&
        products.map(product => (
          <Box key={product.id} margin="0 10px 10px 0">
            <ProductCard product={product} />
          </Box>
        ))}
    </Box>
  );
};

export default SearchResult;
