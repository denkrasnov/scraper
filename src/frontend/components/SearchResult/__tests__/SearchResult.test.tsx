import React from "react";
import { shallow } from "enzyme";

import * as AppContext from "~app/services/ContextProvider";
import { INITIAL_STATE } from "~app/services/fetchProducts/constants";
import SearchResult from "..";

describe("SearchResult", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render products", () => {
    jest.spyOn(AppContext, "useFullContext").mockImplementationOnce(() => [
      {
        ...INITIAL_STATE,
        products: [
          {
            id: "__ID__",
            title: "__TITLE__",
            imageUrl: "__IMAGE_URL__",
            price: "100",
            shop: "__SHOP__"
          }
        ]
      },
      () => {}
    ]);
    const component = shallow(<SearchResult />);

    expect(component).toMatchSnapshot();
  });

  it("should render Loading", () => {
    jest.spyOn(AppContext, "useFullContext").mockImplementationOnce(() => [
      {
        ...INITIAL_STATE,
        isLoading: true
      },
      () => {}
    ]);

    const component = shallow(<SearchResult />);
    expect(component).toMatchSnapshot();
  });

  it("should render Error", () => {
    jest.spyOn(AppContext, "useFullContext").mockImplementationOnce(() => [
      {
        ...INITIAL_STATE,
        isError: "SomeError"
      },
      () => {}
    ]);
    const component = shallow(<SearchResult />);

    expect(component).toMatchSnapshot();
  });
});
