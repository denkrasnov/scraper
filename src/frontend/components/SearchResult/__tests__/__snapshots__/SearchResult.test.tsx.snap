// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult should render Error 1`] = `
<Box
  justifyContent="center"
  marginTop="s48"
>
  Error...
</Box>
`;

exports[`SearchResult should render Loading 1`] = `
<Box
  justifyContent="center"
  marginTop="s48"
>
  <Loader />
</Box>
`;

exports[`SearchResult should render products 1`] = `
<Grid
  result={true}
>
  <Filter
    items={
      Array [
        Object {
          "channel": "TV8",
          "date": "9:00",
          "header": "__TITLE__",
          "imageURL": "__IMAGE_URL__",
          "newsUrl": "__URL__",
        },
      ]
    }
    options={
      Array [
        "NTV",
        "TV8",
        "JurnalTV",
      ]
    }
    productName="MD"
  >
    <ProductList />
  </Filter>
</Grid>
`;

exports[`SearchResult should render when data is undefined 1`] = `
<Box
  justifyContent="center"
  marginTop="s48"
>
  Error...
</Box>
`;
